### PR TITLE
BUG: fix parsing of ODF time values with comments

### PR DESCRIPTION
**Pull request by mzuzek**
- [x] Similar to: #55200 variant for `time-value` cells
- [x] Tests: added comments to cells in `test_1900.ods` and `test_1904.ods` fixtures, so that [`io/excel/test_readers.py:test_reader_seconds()`](https://github.com/pandas-dev/pandas/blob/main/pandas/tests/io/excel/test_readers.py#L976-L1015) would be failing without this fix. Also fixed missing microseconds there (see p.1 in #55045).
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/v2.2.0.rst` file.

> ODF `office:time-value` format is specified by [p.3.2.6 of XML Schema (Part 2)](https://www.w3.org/TR/2004/REC-xmlschema-2-20041028/#duration) as referenced in [p.19.389 of OpenDocument Schema Part 3](https://docs.oasis-open.org/office/OpenDocument/v1.3/OpenDocument-v1.3-part3-schema.html#attribute-office_value-type)

Related to: #55045 (test files updates)

### CODE REVIEW

File: doc/source/whatsnew/v2.2.0.rst [hunk 1]
```diff
@@ -252,6 +252,7 @@ Bug fixes
 - Bug in :class:`pandas.core.window.Rolling` where duplicate datetimelike indexes are treated as consecutive rather than equal with ``closed='left'`` and ``closed='neither'`` (:issue:`20712`)
 - Bug in :meth:`DataFrame.apply` where passing ``raw=True`` ignored ``args`` passed to the applied function (:issue:`55009`)
 - Bug in :meth:`pandas.read_excel` with a ODS file without cached formatted cell for float values (:issue:`55219`)
+- Bug in :meth:`pandas.read_excel` where ODS files with comments on time value cells failed to parse (related to :issue:`55200`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.2.0.rst:255]:
```suggestion
- Bug in :meth:`pandas.read_excel` where ODS files with comments on time value cells failed to parse (:issue:`55200`)
```

File: pandas/io/excel/_odfreader.py [hunk 2]
```diff
@@ -182,6 +183,20 @@ def _get_column_repeat(self, cell) -> int:
...
         return int(cell.attributes.get((TABLENS, "number-columns-repeated"), 1))
 
+    def _parse_odf_time(self, value: str) -> pd.Timestamp:
+        """
+        Helper function to convert ODF variant of ISO 8601 formatted duration
+        "PnYnMnDTnHnMnS" - see https://www.w3.org/TR/xmlschema-2/#duration
+        """
+        parts = re.match(r"^\s*PT\s*(\d+)\s*H\s*(\d+)\s*M\s*(\d+(\.\d+)?)\s*S$", value)
+        if parts is None:
+            raise ValueError(f"Failed to parse ODF time value: {value}")
```
Comment (mroeschke) [inline pandas/io/excel/_odfreader.py:193]: Can you add a test that hits this?

File: pandas/io/excel/_odfreader.py [hunk 3]
```diff
@@ -182,6 +183,20 @@ def _get_column_repeat(self, cell) -> int:
...
+        h, m, s = parts.group(1, 2, 3)
```
Comment (mroeschke) [inline pandas/io/excel/_odfreader.py:194]: Could you avoid 1 variable names here?

File: pandas/io/excel/_odfreader.py [hunk 4]
```diff
@@ -182,6 +183,20 @@ def _get_column_repeat(self, cell) -> int:
...
+        # ignore date part from some representations as both pd.Timestamp
+        # and datetime.time restrict hour values to 0..23
+        h = str(int(h) % 24)
+        return pd.Timestamp(f"{h}:{m}:{s}")
```
Comment (mroeschke) [inline pandas/io/excel/_odfreader.py:198]: Could we just assemble a `datetime.time` object here?

File: pandas/io/excel/_odfreader.py [hunk 5]
```diff
@@ -182,6 +183,20 @@ def _get_column_repeat(self, cell) -> int:
 
         return int(cell.attributes.get((TABLENS, "number-columns-repeated"), 1))
 
+    def _parse_odf_time(self, value: str) -> pd.Timestamp:
+        """
+        Helper function to convert ODF variant of ISO 8601 formatted duration
+        "PnYnMnDTnHnMnS" - see https://www.w3.org/TR/xmlschema-2/#duration
+        """
```
Comment (rhshadrach) [inline pandas/io/excel/_odfreader.py:190]: Nit: NumPy-style docstrings should have a one line summary followed by a blank line. You can include more details after the one line summary.

File: pandas/io/excel/_odfreader.py [hunk 6]
```diff
@@ -182,6 +183,20 @@ def _get_column_repeat(self, cell) -> int:
...
+        parts = re.match(r"^\s*PT\s*(\d+)\s*H\s*(\d+)\s*M\s*(\d+(\.\d+)?)\s*S$", value)
```
Comment (rhshadrach) [inline pandas/io/excel/_odfreader.py:191]: This is going to be called many times, can you use a compiled regex.

**Review by rhshadrach — CHANGES_REQUESTED** [review]

File: pandas/io/excel/_odfreader.py [hunk 7]
```diff
@@ -182,6 +183,20 @@ def _get_column_repeat(self, cell) -> int:
...
+        if parts is None:
+            raise ValueError(f"Failed to parse ODF time value: {value}")
+        h, m, s = parts.group(1, 2, 3)
+        # ignore date part from some representations as both pd.Timestamp
+        # and datetime.time restrict hour values to 0..23
```
Comment (dimastbk) [inline pandas/io/excel/_odfreader.py:196]: ODS supports timedelta (`[hh]:mm:ss`). `times_1904.ods` is broken, see https://github.com/pandas-dev/pandas/issues/55045. Maybe something like?
```python
if h > 23:
    return pd.Timedelta(...)
else:
    return cast(Scalar, datetime.time(...))
```

File: pandas/io/excel/_odfreader.py [hunk 8]
```diff
@@ -182,6 +190,27 @@ def _get_column_repeat(self, cell) -> int:
 
         return int(cell.attributes.get((TABLENS, "number-columns-repeated"), 1))
 
+    def _parse_odf_time(self, value: str) -> datetime.time:
```
Comment (rhshadrach) [inline pandas/io/excel/_odfreader.py:193]: In the original issue, a suggestion is to skip annotations in ODFReader. Why take the approach of parsing times by regex instead?

File: doc/source/whatsnew/v2.2.0.rst (hunk 1 above)
Comment (mzuzek) [reply doc/source/whatsnew/v2.2.0.rst]: Thanks - applied, assuming it's clear enough that while this PR takes care of comments on *time* values #55200 targets *string* values and requires a separate fix

File: pandas/io/excel/_odfreader.py (hunk 2 above)
Comment (mzuzek) [reply pandas/io/excel/_odfreader.py:199]: Sure, on 6a5d4eb8d19dbbbc11811dbdc3a2792a8f5690e3 I've added new sample with manually corrupted time value.

File: pandas/io/excel/_odfreader.py (hunk 3 above)
Comment (mzuzek) [reply pandas/io/excel/_odfreader.py]: I'm sorry, fixed on 0c70cc2a97a38c52922fd0dc6e7553d2a40d7c04

File: pandas/io/excel/_odfreader.py (hunk 4 above)
Comment (mzuzek) [reply pandas/io/excel/_odfreader.py]: Thanks, fixed on 0c70cc2a97a38c52922fd0dc6e7553d2a40d7c04 (crossed my mind, but I hesitated to cut off possible hidden advantages brought by `Timestamp` intermediate)

File: pandas/io/excel/_odfreader.py (hunk 5 above)
Comment (mzuzek) [reply pandas/io/excel/_odfreader.py]: I'm sorry, fixed on 6ebe03a92d45802a84a504b31fd57a492e364ed5

File: pandas/io/excel/_odfreader.py (hunk 6 above)
Comment (mzuzek) [reply pandas/io/excel/_odfreader.py]: Good point, thanks! Fixed on 6ebe03a92d45802a84a504b31fd57a492e364ed5.

File: pandas/io/excel/_odfreader.py (hunk 7 above)
Comment (mzuzek) [reply pandas/io/excel/_odfreader.py]: Right, `timedelta` seem to make more sense than timestamp - at least for ODF `time-value`, because durations can be larger than 24h and they can be negative, see [duration.ods](https://github.com/pandas-dev/pandas/files/12802718/duration.ods):
duration

It's my first PR here, so I thought I'd better keep it tight and clean - hoping we could leave `timedelta` for a follow-up discussion and new PR. For sure unit tests would need adjustments (they specifically require `datetime.time` timestamps) and I don't know how other spreadsheet formats correspond.

File: pandas/tests/io/data/excel/times_1904.ods
Comment (mzuzek) [inline pandas/tests/io/data/excel/times_1904.ods:1]: FYI: In `times_1900.ods` and `times_1904.ods` I've added missing microseconds to cells' `office:time-value` tags, because the tests started failing (nice coverage!) on different value that was previously obtained from cell string representation.

I understand this fixes p.1 of #55045 (thank you @dimastbk for referencing the issue), I've left p.2 intact, because it's legal ODF representation - see https://github.com/pandas-dev/pandas/pull/55324#discussion_r1345649211.

diff

File: pandas/io/excel/_odfreader.py (hunk 8 above)
Comment (mzuzek) [reply pandas/io/excel/_odfreader.py]: I'm not sure if string cells (which #55200 is focused on) have similar pitfall, but time/duration cells have their actual value stored in `office:time-value` and `text:p` only carries "display value" which - affected by cell formatting  - can be different and misleading. That's why I propose to replace original `str(cell)` (merging `text:p` values) with the extraction of `office:time-value` regexp.

For example, see check out [text-repr.ods](https://github.com/pandas-dev/pandas/files/12802867/text-repr.ods):
text-repr

**Comment by mzuzek** [comment]
On a493bb794eee30a48edefc2f920fcb093dfd2276, I've refactored the helper function into `_get_cell_time_value()` - similar to existing `_get_cell_string_value()` for strings.
Pardon me not grasping the `ODFReader` class layout in the first place...

**Comment by mzuzek** [comment]
@rhshadrach @mroeschke @dimastbk

Would there be anything else I should do on this PR ?

**Review by mroeschke — APPROVED** [review]
Looks OK to me

File: pandas/io/excel/_odfreader.py (hunk 8 above)
Comment (rhshadrach) [reply pandas/io/excel/_odfreader.py]: Then this doesn't close #55200 as is indicated in the OP?

**Comment by rhshadrach** [comment]
> [mzuzek](https://github.com/mzuzek) [force-pushed](https://github.com/pandas-dev/pandas/compare/7c22815ea69a3b2e1bbd117b88a90db47ab24ed5..ea23a2e501a8db58b07907899383d3ce3e49faa3) the fix-odf-time-comments branch from [7c22815 ](https://github.com/pandas-dev/pandas/commit/7c22815ea69a3b2e1bbd117b88a90db47ab24ed5)to [ea23a2e ](https://github.com/pandas-dev/pandas/commit/ea23a2e501a8db58b07907899383d3ce3e49faa3)
[yesterday](https://github.com/pandas-dev/pandas/pull/55324#event-10571687912)

Just so you're aware, by force pushing you make it so that reviewers can no longer use the "Show changes since your last review" feature. Not a big deal at all here because the diff is so small.

File: pandas/tests/io/data/excel/times_1904.ods
Comment (rhshadrach) [reply pandas/tests/io/data/excel/times_1904.ods:1]: > FYI: In `times_1900.ods` and `times_1904.ods` I've added missing microseconds to cells' `office:time-value` tags,

and

> values like `PT35089H02M03` in `times_1904.ods` are legal in ODF

Does this mean the ods files in the test suite no longer have a time-value with missing microseconds even though it's legal?

**Review by rhshadrach — CHANGES_REQUESTED** [review]
I have some questions - see the above review comments.

File: pandas/io/excel/_odfreader.py (hunk 8 above)
Comment (mzuzek) [reply pandas/io/excel/_odfreader.py]: Not at all, I tried to indicate in [the description](https://github.com/pandas-dev/pandas/pull/55324#issue-1919280333) that this PR fixes a new bug which is a twin sister of [#55200](https://github.com/pandas-dev/pandas/issues/55200) targeting _time_ cells instead of _string_ cells. Also underlined that [above](https://github.com/pandas-dev/pandas/pull/55324#discussion_r1345532050), so please check if [the release comment](https://github.com/pandas-dev/pandas/pull/55324#discussion_r1341671356) is accurate in its current form. I'm sorry if I failed to make things clear enough.

Comment (mzuzek) [reply pandas/io/excel/_odfreader.py]: I passed on reporting new bug report, because this PR technically servers as one (in addition to description, it contains correct ODF samples in form of updated `*.ods` files that break current tests).
Let me know if a new ticket should have been created for this. Alternatively, you could consider extending the scope of [#55200](https://github.com/pandas-dev/pandas/issues/55200) to cover comments on _all types_ of spreadsheet cells (not only strings).

File: pandas/tests/io/data/excel/times_1904.ods
Comment (mzuzek) [reply pandas/tests/io/data/excel/times_1904.ods:1]: The error was raised by the unit test (see [`io/excel/test_readers.py`](https://github.com/pandas-dev/pandas/blob/cbac7cdecd45e2fc07a0a187b8591bada7e75736/pandas/tests/io/excel/test_readers.py#L993-L1015)) which expects the microseconds but saw the new implementation returning times _without_ them.

New implementation proposed here only reads `time-value` tag and totally ignores text representation, so it won't raise any errors related to display value or inconsistency between actual value and display value.

Probably the only way for the user to legally obtain a time cell displaying microseconds but not actually having them (as in the samples) would be to craft somewhat malicious custom cell formatting like this (see [fake_ms.ods](https://github.com/pandas-dev/pandas/files/12846326/fake_ms.ods)):
fake_ms
There was no such formatting in the samples, so normally when they loaded, office fixed the display values by cutting off the microseconds (took me few moments to figure out the puzzle and find where they hide 😉). This inconsistency in `value + format = display_value` makes the prior samples incorrect, as described in p.1 of [#55045](https://github.com/pandas-dev/pandas/issues/55045).

Would that make sense ?

File: pandas/io/excel/_odfreader.py (hunk 8 above)
Comment (rhshadrach) [reply pandas/io/excel/_odfreader.py]: The issue is that putting `Fix: #...` links the issue to this PR so that upon merging that issue would be closed. I've updated the OP.

File: pandas/tests/io/data/excel/times_1904.ods
Comment (rhshadrach) [reply pandas/tests/io/data/excel/times_1904.ods:1]: Let me try again.

> values like PT35089H02M03 in times_1904.ods are legal in ODF

That value doesn't seem to match the regex `^\s*PT\s*(\d+)\s*H\s*(\d+)\s*M\s*(\d+)(\.(\d+))?\s*S$`. If I am correct here, isn't this an issue?

Comment (mzuzek) [reply pandas/tests/io/data/excel/times_1904.ods:1]: My bad, `PT35089H02M03` (missing trailing `S`) was just a typo… I said there it’s legal „to denote duration of 35089 hours 2m and 3s” (i.e. [more than 24](https://github.com/pandas-dev/pandas/pull/55324#discussion_r1345649211)) and the screenshots above show that `test_1904.ods` contains values with `S` before and after adding microseconds.

File: pandas/io/excel/_odfreader.py (hunk 7 above)
Comment (rhshadrach) [reply pandas/io/excel/_odfreader.py]: Am I correct in saying that the current behavior of pandas will read the first value as having 50 hours, where as this change will now be `50 - 48 = 2` hours?

Comment (mzuzek) [reply pandas/io/excel/_odfreader.py]: @rhshadrach Which commit/version of pandas yields 50 hours for you ?
I thought you'd get an error in [_odfreader.py:217](https://github.com/pandas-dev/pandas/blob/e0d6051f985994e594b07a2b93b9ca2eff43eae4/pandas/io/excel/_odfreader.py#L217) similar to:
```ruby
>>> pd.Timestamp('50:15:00') 
Traceback (most recent call last):
  File "parsing.pyx", line 681, in pandas._libs.tslibs.parsing.dateutil_parse
ValueError: hour must be in 0..23
```

Comment (rhshadrach) [reply pandas/io/excel/_odfreader.py]: > Which commit/version of pandas yields 50 hours for you ?

I haven't run anything.

> I thought you'd get an error in [_odfreader.py:217](https://github.com/pandas-dev/pandas/blob/e0d6051f985994e594b07a2b93b9ca2eff43eae4/pandas/io/excel/_odfreader.py#L217) similar to:

I think you're saying that both main and this PR will raise on `duration.ods`, is that right?

Comment (mzuzek) [reply pandas/io/excel/_odfreader.py]: Nope, I only said that the current implementation does not support time value equal or larger than 24 hours in ODF files.

Could you please run the files you'd like to check and share any results that raise your concerns ?

Comment (rhshadrach) [reply pandas/io/excel/_odfreader.py]: > I only said that the current implementation

Does current implementation mean main or this PR?

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Comment by mzuzek** [comment]
@rhshadrach @mroeschke @dimastbk

Thank you for reviewing this PR and for all your comments.

I'm sorry it didn't work out:
I understand that after all review issues and requests have been properly addressed - as confirmed [here](https://github.com/pandas-dev/pandas/pull/55324#pullrequestreview-1662225869) -    
this PR was blocked by discussions which haven't introduced any new problems or change requests (last one [here](https://github.com/pandas-dev/pandas/pull/55324#discussion_r1353434324)).  
I'll assume it's just the way things work here.
